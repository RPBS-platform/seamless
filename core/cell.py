"""Module containing the Cell class."""

import traceback
import inspect
import ast
import os
import copy

from .. import dtypes
from .utils import find_return_in_scope
from .process import Managed
from . import libmanager
from .resource import Resource

class CellLike(object):
    """Base class for cells and contexts
    CellLikes are captured by context.cells"""
    _like_cell = True

class Cell(Managed, CellLike):
    """Default class for cells.

    Cells contain all the state in text form
    """

    class StatusFlags:
        UNINITIALISED, ERROR, OK = range(3)
    StatusFlagNames = ["UNINITIALISED", "ERROR", "OK"]

    _dtype = None
    _data = None  # data, always in text format
    _data_last = None

    _error_message = None
    _status = StatusFlags.UNINITIALISED

    _dependent = False

    _incoming_connections = 0
    _outgoing_connections = 0

    _resource = None

    def __init__(self, dtype, *, naming_pattern="cell"):
        """TODO: docstring."""
        super().__init__()
        from .macro import get_macro_mode
        from .context import get_active_context
        assert dtypes.check_registered(dtype), dtype
        self._dtype = dtype
        self._last_object = None
        self._resource = Resource(self)
        if get_macro_mode():
            ctx = get_active_context()
            ctx._add_new_cell(self, naming_pattern)

    @property
    def resource(self):
        return self._resource

    @property
    def dependent(self):
        """Indicate if the cell is dependent.

        Property is true if the cell has a hard incoming connection,
        e.g. the output of a process.
        """
        return self._dependent

    def set(self, text_or_object):
        """Update cell data from Python code in the main thread."""
        if isinstance(text_or_object, (str, bytes)):
            self._text_set(text_or_object, trusted=False)
        else:
            self._object_set(text_or_object, trusted=False)
        return self

    def fromfile(self, filename):
        return self.resource.fromfile(filename, frames_back=2)

    def _text_set(self, data, trusted):
        try:
            if self._status == self.__class__.StatusFlags.OK \
                    and (data is self._data or data is self._data_last or
                         data == self._data or data == self._data_last):
                return False
        except:
            pass

        try:
            """Check if we can parse the text"""
            dtypes.parse(self._dtype, data, trusted=trusted)

        except dtypes.ParseError:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise
        else:
            self._data_last = data
            self._data = data
            self._status = self.__class__.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
        return True

    def _object_set(self, object_, trusted):
        if self._status == self.__class__.StatusFlags.OK \
                and object_ == self._last_object:
            return False
        try:
            """
            Construct the object:
             If the object is already of the correct type,
               then constructed_object is object_
             Some datatypes (i.e. silk) can construct the object from
              heterogenous input
            """
            dtypes.construct(self._dtype, object_)

        except dtypes.ConstructionError:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise
        else:
            data = dtypes.serialize(self._dtype, object_)
            # Normally no error here...
            self._data = data
            self._status = self.__class__.StatusFlags.OK
            self._last_object = copy.deepcopy(object_)

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
        return True

    def touch(self):
        if self._status != self.__class__.StatusFlags.OK:
            return
        if self._context is not None:
            manager = self._get_manager()
            manager.update_from_code(self)

    def _update(self, data):
        """Invoked when cell data is updated by a process."""
        return self._text_set(data, trusted=True)

    def connect(self, target):
        """Connect the cell to a process's input pin."""
        manager = self._get_manager()
        manager.connect(self, target)

    @property
    def dtype(self):
        """The cell's data type."""
        return self._dtype

    @property
    def data(self):
        """The cell's data in text format."""
        return copy.deepcopy(self._data)

    @property
    def status(self):
        """The cell's current status."""
        return self.StatusFlagNames[self._status]

    @property
    def error_message(self):
        """The cell's current error message.

        Returns None is there is no error
        """
        return self._error_message

    def _on_connect(self, pin, process, incoming):
        from .process import OutputPinBase
        if incoming:
            if self._dependent and isinstance(pin, OutputPinBase):
                raise Exception(
                 "Cell is already the output of another process"
                )
            if isinstance(pin, OutputPinBase):
                self._dependent = True
            self._incoming_connections += 1
        else:
            self._outgoing_connections += 1

    def _on_disconnect(self, pin, process, incoming):
        from .process import OutputPinBase
        if incoming:
            if isinstance(pin, OutputPinBase):
                self._dependent = False
            self._incoming_connections -= 1
        else:
            self._outgoing_connections -= 1

    def _set_error_state(self, error_message=None):
        if error_message is not None:
            self._status = self.StatusFlags.ERROR
            if error_message != self._error_message:
                print(error_message)
        self._error_message = error_message

    def add_macro_object(self, macro_object, macro_arg):
        manager = self._get_manager()
        manager.add_macro_listener(self, macro_object, macro_arg)

    def remove_macro_object(self, macro_object, macro_arg):
        manager = self._get_manager()
        manager.remove_macro_listener(self, macro_object, macro_arg)

    def destroy(self):
        if self._destroyed:
            return
        #print("CELL DESTROY", self)
        self.resource.destroy()
        super().destroy()

class PythonCell(Cell):
    """
    A cell containing Python code.

    Python cells may contain either a code block or a function
    Processes that are connected to it may require either a code block or a
     function
    Mismatch between the two is not a problem, unless:
          Connected processes have conflicting block/function requirements
        OR
            A function is required (typically, true for transformers)
          AND
            The cell contains a code block
          AND
            The code block contains no return statement
    """

    class CodeTypes:
        ANY, FUNCTION, BLOCK = range(3)

    _dtype = ("text", "python")

    _code_type = CodeTypes.ANY
    _required_code_type = CodeTypes.ANY

    def _text_set(self, data, trusted):
        if data == self._data:
            return False
        try:
            """Check if the code is valid Python syntax"""
            ast_tree = compile(data, str(self), "exec", ast.PyCF_ONLY_AST)

        except SyntaxError:
            if not trusted:
                raise

            else:
                self._set_error_state(traceback.format_exc())

        else:
            is_function = (
             len(ast_tree.body) == 1 and
             isinstance(ast_tree.body[0], ast.FunctionDef)
            )

            # If this cell requires a function, but wasn't provided
            #  with a def block
            if not is_function and \
                    self._required_code_type == self.CodeTypes.FUNCTION:
                # Look for return node in AST
                try:
                    find_return_in_scope(ast_tree)
                except ValueError:
                    exception = SyntaxError(
                     "Block must contain return statement(s)"
                    )

                    if trusted:
                        self._set_error_state("{}: {}".format(
                         exception.__class__.__name__, exception.msg)
                        )
                        return

                    else:
                        raise exception

            self._data = data
            self._code_type = self.CodeTypes.FUNCTION if is_function else \
                self.CodeTypes.BLOCK
            self._set_error_state(None)
            self._status = self.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
            return True

    def _object_set(self, object_, trusted):
        from .utils import strip_source
        try:
            """
            Try to retrieve the source code
            Will only work if code is a function
            """
            if not inspect.isfunction(object_):
                raise Exception("Python object must be a function")

            code = inspect.getsource(object_)
            code = strip_source(code)

        except:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise

        else:
            self._code_type = self.CodeTypes.FUNCTION
            oldcode = self._data
            self._data = code
            self._status = self.__class__.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
            return code != oldcode

    def _on_connect(self, pin, process, incoming):
        exc1 = """Cannot connect to %s: process requires a code function
        whereas other connected processes require a code block"""
        exc2 = """Cannot connect to %s: process requires a code block
        whereas other connected processes require a code function"""

        if not incoming:
            if self._required_code_type == self.CodeTypes.BLOCK and \
                    process._required_code_type == self.CodeTypes.FUNCTION:
                raise Exception(exc1 % type(process))
            elif self._required_code_type == self.CodeTypes.FUNCTION and \
                    process._required_code_type == self.CodeTypes.BLOCK:
                raise Exception(exc2 % type(process))

        Cell._on_connect(self, pin, process, incoming)
        if not incoming:
            self._required_code_type = process._required_code_type

    def _on_disconnect(self, pin, process, incoming):
        Cell._on_disconnect(self, pin, process, incoming)
        if self._outgoing_connections == 0:
            self._required_code_type = self.CodeTypes.ANY

_handlers = {
    ("text", "code", "python"): PythonCell
}


def cell(dtype):
    """Factory function for a Cell object."""
    cell_cls = Cell
    if dtype in _handlers:
        cell_cls = _handlers[dtype]

    newcell = cell_cls(dtype)
    return newcell


def pythoncell():
    """Factory function for a PythonCell object."""
    return cell(("text", "code", "python"))
