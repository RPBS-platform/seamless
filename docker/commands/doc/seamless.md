% seamless(1) Version 0.1 | seamless-cli Documentation

NAME
====

**seamless** - Starts a new Seamless Docker container and runs a command inside it.

SYNOPSIS
========

| **seamless** \[_command_]

DESCRIPTION
===========

Starts a new Seamless Docker container and runs a command inside it.

The current directory is mounted to /cwd, and the command is executed there

/tmp is mounted as well

Uses the host network for the Docker container. Will only work under Linux.

BUGS
====

See GitHub Issues: <https://github.com/sjdv1982/seamless/issues>

AUTHOR
======

Sjoerd de Vries <sjdv1982@gmail.com>

SEE ALSO
========

**seamless-bash(1)**, **seamless-ipython(1)**, **seamless-jupyter(1)**