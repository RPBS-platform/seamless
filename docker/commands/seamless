set -u -e

firstarg=$1 #there must be at least one argument
gdbflags='--cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined'
docker run --rm \
  $gdbflags \
  --network=host \
  -v `pwd`:/cwd \
  -v /tmp:/tmp \
  --workdir /cwd \
  --user $(id -u):$(id -g) \
  rpbs/seamless $*
