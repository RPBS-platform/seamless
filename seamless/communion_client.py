import asyncio

REMOTE_TIMEOUT = 5.0

class CommunionClient:
    pass

class CommunionBufferClient(CommunionClient):
    config_type = "buffer"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_buffer = config["buffer"]
        self.config_status = config["buffer_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "buffer_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        if result is not None:
            try:
                status, result_checksum = result
                if status in (0, 1):
                    communion_client_manager.remote_checksum_available.add(
                        result_checksum
                    )
            except:
                pass
        return result
    
    async def submit(self, checksum):
        assert checksum is not None
        if not self.config_buffer:
            return
        message = {
            "type": "buffer",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result


class CommunionBufferLengthClient(CommunionClient):
    config_type = "buffer_length"

    def __init__(self, servant, config):
        self.servant = servant

    async def submit(self, checksum):
        assert checksum is not None
        message = {
            "type": "buffer_length",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result

class CommunionTransformationClient(CommunionClient):
    config_type = "transformation"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_job = config["transformation_job"]
        self.config_status = config["transformation_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "transformation_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result
    
    async def submit(self, checksum):
        if not self.config_job:
            return
        message = {
            "type": "transformation_job",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result


class CommunionBuildModuleClient(CommunionClient):
    config_type = "build_module"

    def __init__(self, servant, config):
        self.servant = servant
        self.config_job = config["build_module_job"]
        self.config_status = config["build_module_status"]

    async def status(self, checksum):
        assert checksum is not None
        if not self.config_status:
            return
        message = {
            "type": "build_module_status",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result
    
    async def submit(self, checksum):
        if not self.config_job:
            return
        message = {
            "type": "transformation_job",
            "content": checksum
        }
        result = await communion_server.client_submit(message, self.servant)
        return result

"""
class CommunionBuildModuleClient(CommunionClient):
    config_type = "build_module"
    cache_task_servers = remote_build_model_servers
    def _prepare_message(self, content):
        return {
            "type": self.config_type,
            "content": content,
        }

class CommunionTransformerResultClient(CommunionClient):
    config_type = "transformer_result"
    cache_task_servers = remote_transformer_result_servers
    def _prepare_message(self, checksum):
        return {
            "type": self.config_type,
            "content": checksum,
        }


class CommunionTransformerJobClient(CommunionPairClient):
    config_type = "transformer_job"
    cache_task_servers = remote_job_servers
    def _prepare_message(self, funcmode, checksum):
        if funcmode == "check":
            return {
                "type": "transformer_job_check",
                "content": checksum,
            }
        elif funcmode == "run":
            return {
                "type": "transformer_job_run",
                "content": checksum,
            }


"""

class CommunionClientManager:
    _clientclasses = [klass for klass in globals().values() if isinstance(klass, type) \
      and issubclass(klass, CommunionClient) and klass is not CommunionClient]

    _clientclasses = {klass.config_type:klass for klass in _clientclasses}

    def __init__(self):
        self.clients = {k:[] for k in self._clientclasses}
        self.remote_checksum_available = set()
        self.servant_to_clients = {}
        self.servant_to_peer_id = {}

    def add_servant(self, servant, peer_id, *,
            config_servant, config_master
    ):
        servid = id(servant)
        self.servant_to_clients[servid] = []
        self.servant_to_peer_id[servid] = peer_id
        communion_types = {
            "buffer": ["buffer", "buffer_status"],
            "buffer_length": ["buffer_length"],
            "transformation": ["transformation_job", "transformation_status"],
            "build_module": ["build_module_job", "build_module_status"],
        }
        for communion_type in communion_types:
            sub_communion_types = {}
            for sub_communion_type in communion_types[communion_type]:
                c_master = config_master[sub_communion_type]
                c_servant = config_servant[sub_communion_type]                
                if not c_master or not c_servant:
                    c = False
                elif sub_communion_type == "buffer":
                    if c_servant == "small" or c_master == "small":
                        c = "small"
                    else:
                        c = True
                else:
                    c = True
                sub_communion_types[sub_communion_type] = c
            if not any(sub_communion_types.values()):
                continue
            print("ADD SERVANT", communion_type)
            clientclass = self._clientclasses[communion_type]
            client = clientclass(servant, sub_communion_types)
            self.clients[communion_type].append(client)
            self.servant_to_clients[servid].append((communion_type, client))

    def remove_servant(self, servant):
        self.remote_checksum_available.clear()
        clients = self.servant_to_clients.pop(id(servant))
        self.servant_to_peer_id.pop(id(servant))
        for communion_type, client in clients:
            print("REMOVE SERVANT", communion_type)
            self.clients[communion_type].remove(client)

    async def remote_buffer_status(self, checksum, peer_id):
        if checksum in self.remote_checksum_available:
            return True
        clients = []
        for client in self.clients["buffer"]:
            client_peer_id = self.servant_to_peer_id[id(client.servant)]
            if client_peer_id != peer_id:
                clients.append(client)
        if not len(clients):
            return False
        coros = [client.status(checksum) for client in clients]
        futures = [asyncio.ensure_future(coro) for coro in coros]
        while 1:
            done, pending = await asyncio.wait(*futures, 
                timeout=REMOTE_TIMEOUT,
                return_when=asyncio.FIRST_COMPLETED
            )
            for future in done:
                try:
                    result = future.result()
                    if result > 0:
                        return True
                except:
                    continue
            if not len(pending):
                return False         


communion_client_manager = CommunionClientManager()
from .communion_server import communion_server