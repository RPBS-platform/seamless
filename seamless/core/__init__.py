import weakref
from enum import Enum

class IpyString(str):
    def _repr_pretty_(self, p, cycle):
        return p.text(str(self))

class SeamlessBase:
    _destroyed = False
    _macro_object = None # macro object that CREATED this instance
    _context = None
    _last_context = None
    name = None

    StatusFlags = Enum('StatusFlags', ('OK', 'PENDING', 'UNDEFINED', 'UNCONNECTED', 'ERROR'))
    _status = StatusFlags.UNDEFINED

    @property
    def path(self):
        if self._context is None:
            return ()
        elif self._context.path is None:
            return ("<None>", self.name)
        else:
            return self._context.path + (self.name,)

    def _validate_path(self, required_path=None):
        if required_path is None:
            required_path = self.path
        else:
            assert self.path == required_path, (self.path, required_path)
        return required_path


    def _set_context(self, context, name, force_detach=False):
        from .context import Context
        assert isinstance(context, Context)
        if self._context is not None:
            assert self.name is not None
            if context is not self._context or force_detach:
                childname = self.name
                assert self._context._children[childname] is self
                self._context._children.pop(childname)
                if childname in self._context._auto:
                    self._context._auto.remove(childname)
        if context is not None:
            self._last_context = context
        self._context = context
        self.name = name
        return self

    @property
    def context(self):
        return self._context


    def _macro_control(self):
        if self._macro_object is not None:
            return self

        ret = None
        if self.context is not None:
            ret = self.context._macro_control()
        else:
            return None

    def format_path(self):
        if self.path is None:
            ret = "<None>"
        else:
            ret = "." + ".".join(self.path)
        return ret

    def __str__(self):
        ret = "Seamless object: " + self.format_path()
        return ret

    def __repr__(self):
        return self.__str__()

    @property
    def macro(self):
        """Returns the macro object associated with this construct"""
        return self._macro_object
    def _set_macro_object(self, macro_object):
        self._macro_object = macro_object

#from .cell import Cell ###
#from .worker import Worker ###
from .context import Context, context ###
