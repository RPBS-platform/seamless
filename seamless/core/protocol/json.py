import json
from ...mixed import MixedScalar

def seamless_encoder(obj):
    if isinstance(obj, MixedScalar):
        return default_encoder.encode(obj.value)
    else:
        raise TypeError(repr(obj) + " is not JSON serializable")

def json_encode(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
        allow_nan=True, cls=None, indent=None, separators=None,
        default=seamless_encoder, sort_keys=False, **kw):
    """Serialize ``obj`` to a JSON formatted ``str``.

    Wrapper around json.dumps that overrides the default value of "default"
    """
    global default_encoder
    default_encoder = json.JSONEncoder(
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      indent=indent,
      separators=separators,
      default=None,
      sort_keys=sort_keys,
      **kw
    )
    return json.dumps(obj,
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      cls=cls,
      indent=indent,
      separators=separators,
      default=default,
      sort_keys=sort_keys,
      **kw
    )
