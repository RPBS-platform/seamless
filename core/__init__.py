from .cell import cell, pythoncell
from .manager import connect
from .macro import macro
