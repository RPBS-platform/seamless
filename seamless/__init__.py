"""
Seamless: framework for data-driven and live programming
Copyright 2016-2018, Sjoerd de Vries
"""

import sys
import time
import functools
import traceback
import atexit

from abc import abstractmethod
class Wrapper:
    @abstractmethod
    def _unwrap(self):
        pass

from . import silk
from . import mixed

#Dependencies of seamless

# 1. hard dependencies; without these, "import seamless" will fail.
# Still, if necessary, some of these dependencies could be removed, but seamless would have to be more minimalist in loading its lib

import numpy as np

if np.dtype(np.object).itemsize != 8:
    raise ImportError("Seamless requires a 64-bit system")

#silk must be imported before mixed
from . import silk
from . import mixed

from .core import mainloop as _mainloop
from .core.mainloop import mainloop, mainloop_one_iteration, asyncio_finish, workqueue
atexit.register(asyncio_finish)

ipython = None
ipy_error = None
try:
    from IPython import get_ipython
    from IPython.core.error import UsageError
    from IPython.terminal.pt_inputhooks import register as _register_integration_terminal
    from ipykernel.eventloops import register_integration as _register_integration_kernel
except ImportError:
    raise
    ipy_error = "Cannot find IPython"
else:
    ipython = get_ipython()
    if ipython is None:
        ipy_error = "Seamless was not imported inside IPython"

if ipy_error is None:
    last_exception = None
    def new_except_hook(etype, evalue, tb):
        global last_exception
        exc = traceback.format_exception(etype, evalue, tb)
        if exc != last_exception:
            last_exception = exc
            print("".join(exc))

    def patch_excepthook():
        sys.excepthook = new_except_hook
    patch_excepthook()

    def mainloop():
        raise RuntimeError("Cannot run seamless.mainloop() in IPython mode")

    def inputhook_terminal(context):
        while not context.input_is_ready():
            mainloop_one_iteration()
    _register_integration_terminal("seamless", inputhook_terminal)

    @_register_integration_kernel('seamless')
    def inputhook_kernel(kernel):
        while 1:
            t = time.time()
            while time.time() - t < kernel._poll_interval:
                mainloop_one_iteration()        
            kernel.do_one_iteration()

    _register_integration_terminal("seamless", inputhook_terminal)
    
    ipython.enable_gui("seamless")
    
else:
    sys.stderr.write("    " + ipy_error + "\n")
    sys.stderr.write("    Call seamless.mainloop(), seamless.flush() or context.equilibrate() to process cell updates\n")

def flush():
    from .core.mainloop import workqueue
    workqueue.flush()

from .highlevel import *
from .silk import Silk
from .debugger import pdb
