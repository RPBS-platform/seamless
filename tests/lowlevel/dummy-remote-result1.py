"""Define four servers that can perform dummy remote caching for transformer results
The first one returns first, but returns -1 (cache miss)
The second one raises an Exception after 2 secs
The third one returns 42 after 3 seconds
The fourth one returns 43 after 5 seconds, i.e. it should be canceled
"""

import asyncio
import seamless
import json
from seamless.core import context, cell, transformer, macro_mode_on
from seamless import get_hash

def h(value):
    return get_hash(json.dumps(value)+"\n")

async def server1(checksum):
    print("Server 1")
    return -3, None

async def server2(checksum):
    print("Server 2...")
    await asyncio.sleep(2)
    print("... server 2")
    raise Exception # Server 2 raises an exception
    
async def server3(checksum):
    print("Server 3...")
    await asyncio.sleep(3)
    print("... server 3")
    return 3, h(42)

async def server4(checksum):
    print("Server 4...")
    try:
        await asyncio.sleep(5)
        print("... server 4")
        return 3, h(43)
    except asyncio.CancelledError:
        print("... server 4 CANCELLED")
        raise

seamless.set_ncores(0)

class DummyClient:
    def __init__(self, func):
        self.status = func


from seamless.communion_client import communion_client_manager
m = communion_client_manager
m.clients["transformation"] = [DummyClient(s) 
    for s in (server1, server2, server3, server4)
] # dirty hack

with macro_mode_on():
    ctx = context(toplevel=True)
    ctx.cell1 = cell().set(2)
    ctx.cell2 = cell().set(3)
    ctx.code = cell("python").set("c = a + b")
    ctx.result = cell()
    ctx.tf = transformer({
        "a": "input",
        "b": "input",
        "c": "output"
    })
    ctx.cell1.connect(ctx.tf.a)
    ctx.cell2.connect(ctx.tf.b)
    ctx.tf.c.connect(ctx.result)
    ctx.code.connect(ctx.tf.code)
    ctx.hashcell = cell()
    
ctx.hashcell.set(42)
ctx.equilibrate(0.1)
ctx.hashcell.set(43)
ctx.equilibrate(0.1)
ctx.hashcell.set(99)
ctx.equilibrate(0.1)

ctx.equilibrate()
print(ctx.status)
print(ctx.result.value)