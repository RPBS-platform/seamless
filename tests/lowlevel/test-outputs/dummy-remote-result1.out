Server 1
Server 2...
Server 3...
Server 4...
... server 2
Waiting for: Seamless transformer: .tf
... server 3
... server 4 CANCELLED
Status: OK
42
